docs/design/interceptors.md

```diff
@@ -0,0 +1,691 @@
+to be `Any` in practice. This is because interceptors can be registered at the *client* level. Such an interceptor would
+be executed for every operation and would not be able to know the specific input or output type at compile time. 
+Interceptors that execute for any operation are common and expected (e.g. add a header to every outgoing request). 
+The downside to this flexibility is that customers will have to cast the input or output type from the context to the 
+specific operation type. 
+  * NOTE: The modeled operation name will be available in the context (it is already available in the execution context)
+  which can be used on interceptors registered at the client level that want to deal with specific operations.
+
+
+### HTTP Interceptor
+
+
+The `Interceptor` interface defined above would not be useful to a customer without specific types. Each protocol will
+get it's own `typealias` that fills in the generics that make sense for a particular protocol. For instance HTTP based 
+protocol clients would be generated to reference the more specific interceptor type:
+
+```kotlin
+typealias HttpInterceptor = Interceptor<Any, Any, HttpRequest, HttpResponse>
+```
+
+
+#### Alternative 1
+
+Alternatively the `Interceptor` interface could hard code `Any` for `Input` and `Output` types. This would simplify 
+some of the runtime type definitions by removing the propagation of those generics. The end result to the customer would
+be the same (a protocol specific interceptor `typealias` is still expected). 
+
+
+## Interceptor Registration
+
+Interceptors will be able to be registered at client creation time as well as through runtime plugins (TBD).
+
+
+Generated client config will be modified to include a list of interceptors that can be appended to:
+
+```kotlin
+
+public interface FooClient : SdkClient {
+  public class Config private constructor(builder: Builder) {
+    val interceptors: List<HttpInterceptor> = builder.interceptors
+
+    public class Builder {
+        var interceptors: MutableList<HttpInterceptor> = mutableListOf()
+    }
+  }
+} 
+
+```
+
+Usage
+
+```kotlin
+
+val client = FooClient {
+  interceptors += CustomInterceptor1()
+  interceptors += CustomInterceptor2()
+}
+
+class CustomInterceptor1 : HttpInterceptor { ... }
+class CustomInterceptor2 : HttpInterceptor { ... }
+```
+
+
+## Interceptor Priority
+
+Interceptors may be registered directly or via runtime plugins (TBD). Interceptors will be registered and executed 
+in a deterministic order:
+
+* Interceptors registered via. smithy default plugins
+* (AWS Services) Interceptors registered via AWS default plugins (AWS SDKs only)
+* Interceptors registered via. service-customization plugins
+* Interceptors registered via. client-level plugins
+* Interceptors registered via. client-level configuration
+* Interceptors registered via. operation-level plugins
+* Interceptors registered via operation-level configuration
```

Comment:
**Style**: Repeating "Interceptors registered via" on every item is unnecessary. Also, this may be clearer as a numbered (rather than bulleted) list.